optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -121,66 +125,268 @@
      * @param otherClass never null
      * @param joiner never null
      * @param <B> the type of the second matched fact
-     * @return a stream that matches every combination of A and B for which the {@link BiJoiner} is true
+     * @return never null, a stream that matches every combination of A and B for which the {@link BiJoiner} is true
      */
     default <B> BiConstraintStream<A, B> join(Class<B> otherClass, BiJoiner<A, B> joiner) {
-        return join(getConstraintFactory().from(otherClass), joiner);
+        if (joiner instanceof FilteringBiJoiner) {
+            FilteringBiJoiner<A, B> filteringJoiner = (FilteringBiJoiner<A, B>) joiner;
+            return join(otherClass)
+                    .filter(filteringJoiner.getFilter());
+        } else {
+            return join(getConstraintFactory().from(otherClass), joiner);
```

Comment:
Why do we need to handle the FilteringJoiner here already?
The join(UniStream, BiJoiner) method must also deal with wether or not that BiJoiner is a FilteringBiJoiner?